project/JarExtract.scala

```diff
@@ -0,0 +1,157 @@
+    * @param outputJarFile Destination of the thin output jar
+    * @param visitor
+    * @param logger
+    * @return List of extracted files to the `extractedFilesDir`.
+    */
+  def extract(
+    jarFile: Path,
+    extractedFilesDir: Path,
+    outputJarFile: Path,
+    visitor: JarVisitor,
+    logger: Logger
+  ): Seq[File] = {
+    val extractedFiles = scala.collection.mutable.ListBuffer.empty[File]
+    Using(new JarFile(jarFile.toFile)) { inputJar =>
+      Using(new JarOutputStream(Files.newOutputStream(outputJarFile))) {
+        outputJar =>
+          val entries = inputJar.entries()
+          while (entries.hasMoreElements) {
+            val entry     = entries.nextElement()
+            val entryName = entry.getName
+            if (!entry.isDirectory) {
+              visitor.visitJarEntry(entry) match {
+                case Skip =>
+                  logger.debug(s"Skipping entry: ${entryName}")
+                case CopyToOutputJar =>
+                  copyEntry(outputJar, inputJar, entry, logger)
+                case CopyToExtractDir(rename) =>
+                  val destPath = rename match {
+                    case Some(newName) => extractedFilesDir.resolve(newName)
+                    case None          => extractedFilesDir.resolve(entryName)
+                  }
+                  copyEntry(destPath, inputJar, entry, logger)
+                  extractedFiles.append(destPath.toFile)
+              }
+            }
+          }
+      }.recover({ case e: IOException =>
+        logger.err(
+          s"Failed to write to output JAR file at $outputJarFile: ${e.getMessage}"
+        )
+      })
+    }.recover({ case e: IOException =>
+      logger.err(
+        s"Failed to open JAR file at $jarFile: ${e.getMessage}"
+      )
+    })
+    extractedFiles
+  }
+
+  private def copyEntry(
+    dest: JarOutputStream,
+    src: JarFile,
+    entry: JarEntry,
+    logger: Logger
+  ): Unit = {
+    dest.putNextEntry(entry)
+    Using(src.getInputStream(entry)) { is =>
+      is.transferTo(dest)
+    }.recover({ case e: IOException =>
+      logger.error(
+        s"Failed to copy entry ${entry.getName} from JAR: ${e.getMessage}"
+      )
+      e.printStackTrace(System.err)
+    })
+    dest.closeEntry()
+    logger.debug(
+      s"Copied entry ${entry.getName} to output JAR."
+    )
+  }
+
+  private def copyEntry(
+    dest: Path,
+    src: JarFile,
+    entry: JarEntry,
+    logger: Logger
+  ): Unit = {
+    if (!dest.getParent.toFile.exists) {
+      Files.createDirectories(dest.getParent)
+    }
+    Using(src.getInputStream(entry)) { is =>
+      Files.copy(is, dest)
+    }.recover({ case e: IOException =>
+      logger.error(
+        s"Failed to copy entry ${entry.getName} to $dest: ${e.getMessage}"
+      )
+      e.printStackTrace(System.err)
+    })
+    logger.debug(
+      s"Copied entry ${entry.getName} to extractFileDir '$dest'."
+    )
+  }
+
+  val openCVExtractor: JarVisitor = (entry: JarEntry) => {
```

Comment:
- it'd be better if this logic was part of  `build.sbt`
   - people don't want to read code of plugins
   - just configure `<plugin><configure>...</configure></plugin>` in Maven terminology 
- but in some more simplified, declarative form
- maybe some combination of regexps...

Reply:
I believe that the declaration implemented in f9b497e3d1dea22b2f0c7ced6c90d268c4a6052c is better.